### PR TITLE
Migrated to version 3.0.0.0

### PR DESCRIPTION
**Pull request by Tmarndt1**

### CODE REVIEW

File: NetWorkflow.Extensions/NetWorkflow.Extensions.csproj [hunk 1]
```diff
@@ -11,7 +11,7 @@
 		<RepositoryUrl>https://github.com/Tmarndt1/NetWorkflow</RepositoryUrl>
 		<PackageLicenseFile>LICENSE</PackageLicenseFile>
 		<PackageTags>Fluent;NetWorkflow;Net;Workflow;Extensions;</PackageTags>
-		<Version>2.0.0</Version>
+		<Version>2.1.0</Version>
```
Comment (chriswirz) [inline NetWorkflow.Extensions/NetWorkflow.Extensions.csproj]: Don't you mean to make this version 3.0.0?

File: NetWorkflow.Tests/Examples/ParallelWorkflow.cs [hunk 2]
```diff
@@ -66,11 +66,11 @@ public Task<string> RunAsync(Guid args, CancellationToken token = default)
...
         }
 
-        private class Step3 : IWorkflowStepAsync<string[], string>
+        private class Step3 : IWorkflowStepAsync<IEnumerable<string>, string>
         {
-            public Task<string> RunAsync(string[] args, CancellationToken token = default)
+            public Task<string> RunAsync(IEnumerable<string> args, CancellationToken token = default)
             {
-                if (args.Length < 1) return Task.FromResult(string.Empty);
+                if (args.Count() < 1) return Task.FromResult(string.Empty);
```
Comment (chriswirz) [inline NetWorkflow.Tests/Examples/ParallelWorkflow.cs:73]: The former was more efficient. `Count()` is often calculated. Generally speaking, if you can use arrays, they will be faster than other `IEnumerable` types without getting into `unsafe`/`unmanaged` types.

File: NetWorkflow.Tests/WorkflowScheduler_Tests.cs [hunk 3]
```diff
@@ -78,15 +80,17 @@ public void Frequency_Success()
...
                 });
 
+            CancellationTokenSource tokenSource = new CancellationTokenSource();
+
             // Act
-            scheduler.StartAsync();
+            scheduler.StartAsync(tokenSource.Token);
 
             // Assert
             while (true)
```
Comment (chriswirz) [inline NetWorkflow.Tests/WorkflowScheduler_Tests.cs]: Why have a `while (true)` and break on the first pass?

File: NetWorkflow.Tests/WorkflowScheduler_Tests.cs [hunk 4]
```diff
@@ -171,5 +179,35 @@ public void AddWorkflowScheduler_Extensions_Success()
...
+                }))
+                .Configure(config =>
+                {
+                    config.ExecuteAt = WorkflowTime.AtFrequency(TimeSpan.FromMilliseconds(50));
+                });
+
+            CancellationTokenSource tokenSource = new CancellationTokenSource();
+
+            // Act
+            using (scheduler)
```
Comment (chriswirz) [inline NetWorkflow.Tests/WorkflowScheduler_Tests.cs]: Why wouldn't the `using` scope begin with the lifecycle  of that object?

File: NetWorkflow/Executors/WorkflowConditionalExecutor.cs [hunk 5]
```diff
@@ -1,39 +1,45 @@
...
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Threading;
 
 namespace NetWorkflow
 {
-    public class WorkflowExecutorConditional<TIn> : IWorkflowExecutor<TIn, object>
+    internal class WorkflowExecutorConditional<TIn> : IWorkflowExecutor<TIn, object>
```
Comment (chriswirz) [inline NetWorkflow/Executors/WorkflowConditionalExecutor.cs:9]: Why is this no longer part of the public API? You will definitely have up major versions now...

File: NetWorkflow/Executors/WorkflowSkipExecutor.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
+﻿
+using System;
+using System.Threading;
+
+namespace NetWorkflow
+{
+    internal class WorkflowSkipExecutor<TArgs> : IWorkflowExecutor<TArgs, TArgs>
```
Comment (chriswirz) [inline NetWorkflow/Executors/WorkflowSkipExecutor.cs]: if everything is `TArgs` it can just be `T`

File: NetWorkflow/NetWorkflow.csproj [hunk 7]
```diff
@@ -12,7 +12,7 @@
         <RepositoryUrl>https://github.com/Tmarndt1/NetWorkflow</RepositoryUrl>
         <PackageTags>Fluent;NetWorkflow;Net;Workflow;</PackageTags>
         <PackageLicenseFile>LICENSE</PackageLicenseFile>
-		<Version>2.0.0</Version>
+		<Version>3.0.0</Version>
```
Comment (chriswirz) [inline NetWorkflow/NetWorkflow.csproj:15]: Maybe change the commit message to say `3.0.0`

File: NetWorkflow/Scheduler/WorkflowScheduler.cs [hunk 8]
```diff
@@ -65,46 +66,46 @@ public Task StartAsync()
...
                             {
-                                await (Task)_executingMethodAsync.Invoke(_workflowFactory.Invoke(), new object[] { _tokenSource.Token });
+                                await (Task)_executingMethodAsync.Invoke(_workflowFactory.Invoke(), new object[] { token });
                             }
                             else
                             {
-                                _executingMethod.Invoke(_workflowFactory.Invoke(), new object[] { _tokenSource.Token });
+                                _executingMethod.Invoke(_workflowFactory.Invoke(), new object[] { token });
                             }
                         });
```
Comment (chriswirz) [inline NetWorkflow/Scheduler/WorkflowScheduler.cs]: You can give a `ContinueWith` a `CancellationToken` as a second parameter.

File: NetWorkflow/Workflow.cs [hunk 9]
```diff
@@ -79,5 +90,28 @@ public Task<WorkflowResult<TOut>> RunAsync(CancellationToken token = default)
...
             return Task.Run(() => Run(token), token);
         }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!_disposedValue)
+            {
+                if (disposing)
+                {
+                    _next.Dispose();
```
Comment (chriswirz) [inline NetWorkflow/Workflow.cs:100]: `_next?.Dispose();`

**Review by chriswirz — APPROVED** [review]
Unsure why you're moving a lot of your API surface to be internal in `3.0.0`. Otherwise, there is nothing breaking that would stop this PR. That said, if you do accept any of my proposed changes, I can re-review/approve.

File: NetWorkflow/Executors/WorkflowConditionalExecutor.cs (hunk 5 above)
Comment (Tmarndt1) [reply NetWorkflow/Executors/WorkflowConditionalExecutor.cs:9]: Should have been done from the beginning. Only thing returned to end user are public Builder interfaces. Want to hide the implementations. Thoughts?

File: NetWorkflow/NetWorkflow.csproj (hunk 7 above)
Comment (Tmarndt1) [reply NetWorkflow/NetWorkflow.csproj:15]: I will when I squash and merge.

File: NetWorkflow.Tests/WorkflowScheduler_Tests.cs (hunk 4 above)
Comment (Tmarndt1) [reply NetWorkflow.Tests/WorkflowScheduler_Tests.cs]: I'll just change to call Dispose but effectively it gets compiled to that.

**Pull request merged**